design/026-requirements_traits.md

```diff
@@ -0,0 +1,125 @@
+
+Package recipes will keep defining their direct dependencies in the same way they were doing in Conan 1.X, but with the possibility of specifying more traits (might not be necessary in many cases, the defaults should be fine). But recipes will have access, once the graph is evaluated, to the ``requires`` of all their dependencies, both direct(declared) and indirect ones (transitive), describing the relation they have with those dependencies with these traits.
+
+
+## Motivation
+
+Conan's main declaration of dependencies is the ``requires`` relationship. Mostly 2 variants of it are also available in Conan 1.X: ``build_requires`` that aim to express a dependency to a build tool, like cmake, and ``private`` which is discouraged for most cases.
+
+But this model does not provide enough information to describe some specifics of the C/C++ compilation model, for example in the general case, when one library is linking another library, the consumers of the former shouldn’t have visibility over the headers of the later one. A ``requires`` trait called ``transitive_headers`` could control this visibility.
+It seems the potential combinations of dependencies usage is too high to consider approaches like the current one, defining high level names as ``build_requires``, ``test_requires``, ``requires_shared``, etc. does not scale.
+
+
+## Detailed Design
+
+The proposed traits are arguments to the ``requires``, like:
+
+```python
+def requirements(self):
+    self.requires(ref, headers=True, libs=True, run=None, visible=True, …)
+```
+
+There are defaults defined, so the same previous Conan 1.X syntax of class attribute ``requires ="ref"`` and ``self.requires(ref)`` is still possible.
+
+Reminder, following the GNU standard naming:
+
+- Build context: The machine in which the current Conan process and build is running. If we are cross-compiling in a Windows box for an app that will run in a Linux RaspberryPI environment, the build context is the Windows machine.
+- Host context: The machine that will run the apps and libraries being built, in this example the host context is the Linux running in the RaspberryPI
+
+
+The traits are:
+
+- ``headers`` (default True): This trait indicates that there are headers that are going to be #included from this package at compile time. The dependency will be in the host context.
+- ``libs`` (default True): The dependency contains some library or artifact that will be used at link time of the consumer. The dependency will be in the host context. This trait will be true for direct shared and static libraries, but could be false for indirect static libraries that are consumed via a shared library.
+- ``build`` (default False): This dependency is a build tool, an application or executable, like cmake, that is used exclusively at build time, it is not linked/embedded into binaries, and will be in the build context.
```

Comment:
So - is it an error to have `build` True while `libs` and/or `headers` true?

Reply:
Good question, maybe not, for some special cases, like a compiler, lets say MinGW, that is a build_require, for the "build" context, with gcc executables (``build=True``), but it will also contain some headers and libraries associated to the compiler. In a first step, I wouldn't constraint or check incompatibility on traits, but work towards the positive cases: Is there a package out there that would be in the build context but still require to link with it? Should it be the current dual "protobuf in the build context for protoc, and protobuf in the host context for protobuf library"?

Reply:
Ok then I would reword:
```
 This dependency is a build tool, an application or executable, like cmake, that is used exclusively at build time, it is not linked/embedded into binaries, and will be in the build context.
```
which makes it sound mutually exclusive.

Reply:
Sure, lets clarify that in the doc.

Reply:
> In a first step, I wouldn't constraint or check incompatibility on traits, but work towards the positive cases: Is there a package out there that would be in the build context but still require to link with it?

We are using conan to package simulink sfunctions. They have a "mex" part, i.e. a library required in the build context, as well as a library to be linked in the host context. With conan 1.x we didn't find a good way to model this, yet. Hopefully conan 2.x with this proposal will help us here.